docs/source/user-guide/concepts/expressions-and-contexts.md

```diff
@@ -0,0 +1,192 @@
+
+```python
+--8<-- "python/user-guide/concepts/expressions.py:print-expr"
+```
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:expression"
+--8<-- "python/user-guide/concepts/expressions.py:print-expr"
+```
+
+The output above shows that Polars creates an intermediate representation of your expressions.
+This intermediate representation can then be parsed and optimized by the Polars query engine when it is time to actually run the expression.
+
+## Contexts
+
+Polars expressions need a _context_ in which they are executed to produce a result.
+Depending on the context it is used in, the same Polars expression can produce different results.
+In this section, we will learn about the four simplest contexts that Polars provides[^1]:
+
+1. `select`
+2. `with_columns`
+3. `filter`
+4. `group_by`
+
+We use the dataframe below to show how each of the contexts works.
+
+{{code_block('user-guide/concepts/expressions','df',[])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:df"
+```
+
+### `select`
+
+The selection context `select` applies expressions over columns.
+The context `select` may produce new columns that are aggregations, combinations of other columns, or literals:
+
+{{code_block('user-guide/concepts/expressions','select-1',['select'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:select-1"
+```
+
+The expressions in a context `select` must produce series that are all the same length or have length 1.
+Series that have length 1 will be broadcast to match the length of the remaining series.
+Literals, like the number used above, are treated as series of length 1 and are also broadcast.
+
+Note that broadcasting can also occur within expressions.
+For instance, consider the expression below:
+
+{{code_block('user-guide/concepts/expressions','select-2',['select'])}}
+
+```python exec="on" result="text" session="user-guide/concepts/expressions-and-contexts"
+--8<-- "python/user-guide/concepts/expressions.py:select-2"
+```
+
+Both the subtraction and the division use broadcasting within the expression because the subexpressions that compute the mean and the standard deviation evaluate to single values.
+
+The context `select` is very flexible and powerful and allows you to evaluate arbitrary expressions independent of, and in parallel to, each other.
+This is also true of the other contexts that we will see next.
+
+### `with_columns`
+
+The context `with_columns` is very similar to the context `select`.
+The main difference between the two is that the context `with_columns` retains the original columns and adds new ones according to its input expressions, whereas the context `select` drops the original columns:
```

Comment:
> whereas the context `select` drops the original columns:

This is the wrong mental model. We don't drop any columns. You create a new `DataFrame` with (only) the selected columns. The old `DataFrame` still has all the columns.

This also explains the length differents/requirements between `with_columns` and `select`. 

`select`  creates a new `DataFrame` and the output of `select` dictates the new length, whereas the output of `with_columns` is added together with the original columns of the original dataframe to create a new dataframe with all the new columns, and thus the length of the new columns must match the original length.

```python

df = pl.DataFrame({
    "a": pl.arange(10, eager=True),
    "b": pl.arange(10, eager=True)
})

# this is allowed as we create a new dataframe from only the new columns.
df.select(
    pl.all().head(5)
)

# this fails as the length of the new columns are wrong.
df.with_columns(
    pl.all().head(5).name.suffix("_head")
)
```